# Data Analysis using Python and Pandas

#### Pandas is a flexible and easy-to-use data analysis library built on top of the Python programming language

In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [2]:
# The dataset we are using has been downloaded from https://covid19.org
# This is time-series data representing covid19 status in India
# Read the CSV file using pandas

covid_df = pd.read_csv('./Data/covid19_time_series_India.csv')

##### Data from the file is read and stored in 'DataFrame' obect - one of the core structures in Pandas for storing and working with tabular data

In [3]:
type(covid_df)

pandas.core.frame.DataFrame

In [4]:
# Display the data
covid_df

Date    Date_YMD  Daily Confirmed  Total Confirmed  \
0     30 January   2020-01-30                1                1   
1     31 January   2020-01-31                0                1   
2    01 February   2020-02-01                0                1   
3    02 February   2020-02-02                1                2   
4    03 February   2020-02-03                1                3   
..            ...         ...              ...              ...   
336  31 December   2020-12-31            19026         10286234   
337   01 January   2021-01-01            20159         10306393   
338   02 January   2021-01-02            18144         10324537   
339   03 January   2021-01-03            16678         10341215   
340   04 January   2021-01-04            16278         10357493   

     Daily Recovered  Total Recovered  Daily Deceased  Total Deceased  
0                  0                0               0               0  
1                  0                0               0               0  
2                  0                0               0               0  
3                  0                0               0               0  
4                  0                0               0               0  
..               ...              ...             ...             ...  
336            21969          9881565             244          148427  
337            23838          9905403             237          148664  
338            20903          9926306             216          148880  
339            19658          9945964             215          149095  
340            29209          9975173             200          149295  

[341 rows x 8 columns]

Here's what we can say about the data:
    - The file provides 6 day-wise metrics for Covid 19 in India
    - The metrics are: Daily Confirmed, Total Confirmed, Daily Recovered, Total Recovered, Daily Deceased and Total Deceased
    - Data provided is for 341 days, from 30th January 2020 to 4th January 2021

NOTE: These are officially provided numbers. Actual numbers may vary since not all cases are diagnosed / recorded

In [5]:
# View basic information about the data frame
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341 entries, 0 to 340
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Date             341 non-null    object
 1   Date_YMD         341 non-null    object
 2   Daily Confirmed  341 non-null    int64 
 3   Total Confirmed  341 non-null    int64 
 4   Daily Recovered  341 non-null    int64 
 5   Total Recovered  341 non-null    int64 
 6   Daily Deceased   341 non-null    int64 
 7   Total Deceased   341 non-null    int64 
dtypes: int64(6), object(2)
memory usage: 21.4+ KB


We see that each column contains values of a specific data type. Statistical information can be viewed for numerical columns
(Mean, Standard Deviation, Min / Max Values, and number of null / empty values) using the 'describe' method

In [6]:
covid_df.describe()

Daily Confirmed  Total Confirmed  Daily Recovered  Total Recovered  \
count       341.000000     3.410000e+02       341.000000     3.410000e+02   
mean      30373.879765     3.300570e+06     29252.706745     2.905230e+06   
std       29393.037881     3.791066e+06     29373.625649     3.556720e+06   
min           0.000000     1.000000e+00         0.000000     0.000000e+00   
25%        1580.000000     2.444800e+04       580.000000     5.496000e+03   
50%       22718.000000     1.077873e+06     20977.000000     6.776620e+05   
75%       50488.000000     7.119309e+06     54133.000000     6.146402e+06   
max       97860.000000     1.035749e+07    102070.000000     9.975173e+06   

       Daily Deceased  Total Deceased  
count      341.000000      341.000000  
mean       437.815249    51925.797654  
std        383.847271    54722.137416  
min          0.000000        0.000000  
25%         53.000000      781.000000  
50%        394.000000    26830.000000  
75%        710.000000   108598.000000  
max       2004.000000   149295.000000

We can view a list of all columns in the data frame

In [7]:
covid_df.columns

Index(['Date', 'Date_YMD', 'Daily Confirmed', 'Total Confirmed',
       'Daily Recovered', 'Total Recovered', 'Daily Deceased',
       'Total Deceased'],
      dtype='object')

The number of rows and columns can be viewed using the 'shape' method

In [8]:
covid_df.shape

(341, 8)

Here's a summary of the methods and tools we have looked at so far:

    - pd.read_csv(): reads data from a csv file into a pandas dataframe object
    - .info(): gives basic information on rows, columns and data types
    - .describe(): shows statistical information for numerical columns
    - .columns: gives a list a columns
    - .shape: gives the number of rows and columns 


### Retrieving Data from a Data Frame

Usually, the first thing we do is to extract data from a data frame, e.g. a list of values from a particular column, the counts of a specific parameter and so on. To do this, we need to understand how data is internally represented in a data frame. 

Conceptually, a data frame is a dictionary of lists (and not a list of dictionaries) where the keys are column names, and the values are lists/arrays containing data for the respective columns

In [9]:
# Pandas format is like the following: (dictionary of lists)

covid_data_dict = {
    
    'date': ['2020-08-30', '2020-08-31', '2020-09-01'],
    'daily_confirmed': [2000, 1400, 1565],
    'daily_recovered': [3000, 2500, 4325],
    'daily_deceased': [100, 97, 113]
}

Some benefits of representing data in the above format:
    - All values in a column typically have the same type of value, so it's more efficient to store them in a single array
    - Retrieving the values for a particular row simply requires extracting the elements at a given index from each column 
      array
    - The representation is more compact (column names are recorded only once) compared to other formats that use a 
      dictionary for each row of data (example below)

In [12]:
# Pandas format is NOT as follows: (list of dictionaries)
covid_data_list = [
    
    {'date': '2020-08-30', 'daily_confirmed': 2000, 'daily_recovered': 3000, 'daily_deceased': 100},
    {'date': '2020-08-31', 'daily_confirmed': 1400, 'daily_recovered': 2500, 'daily_deceased': 97},
    {'date': '2020-09-01', 'daily_confirmed': 1565, 'daily_recovered': 4325, 'daily_deceased': 113}
]

With this understanding, we can now guess how to retrieve data from an existing data frame. For example, we can get a list
of values from a specific column using the [ ] index notation

In [13]:
covid_data_dict['daily_recovered']

[3000, 2500, 4325]

In [16]:
# Let's retrieve data for a column from our main data frame
covid_df['Daily Recovered']

0          0
1          0
2          0
3          0
4          0
       ...  
336    21969
337    23838
338    20903
339    19658
340    29209
Name: Daily Recovered, Length: 341, dtype: int64

Each column is represented using a data structure called 'Series', which is essentially a numpy array with some additional
methods and proeprties

In [17]:
type(covid_df['Daily Recovered'])

pandas.core.series.Series

In [18]:
# For simplicity and consistency, let's rename the columns of our main data frame
covid_df.columns = ['date', 'date_ymd', 'daily_confirmed', 'total_confirmed', 'daily_recovered', 'total_recovered',
                   'daily_deceased', 'total_deceased']

To view the first 5 records from the data frame, we can use the 'Head' method. That way, we can also verify that column
renaming was successful

In [20]:
covid_df.head()

date    date_ymd  daily_confirmed  total_confirmed  \
0   30 January   2020-01-30                1                1   
1   31 January   2020-01-31                0                1   
2  01 February   2020-02-01                0                1   
3  02 February   2020-02-02                1                2   
4  03 February   2020-02-03                1                3   

   daily_recovered  total_recovered  daily_deceased  total_deceased  
0                0                0               0               0  
1                0                0               0               0  
2                0                0               0               0  
3                0                0               0               0  
4                0                0               0               0

We had noted above that a column in the data frame is basically a numpy array with additional methods. Like arrays, we can 
retrieve a specific value in the series using the index [ ] notation

In [21]:
covid_df['daily_recovered'][250]

81945

Pandas also provides the 'at' method to retrieve the element at a specific row and column directly

In [22]:
covid_df.at[250, 'daily_recovered']

81945

Apart from the index notation, Pandas also allows accessing columns using the '.' notation. However, note that this property works only when the column has no spaces or special characters

In [23]:
covid_df.daily_recovered

0          0
1          0
2          0
3          0
4          0
       ...  
336    21969
337    23838
338    20903
339    19658
340    29209
Name: daily_recovered, Length: 341, dtype: int64

Also, you can pass a list of columns with the index [ ] notation to access a subset of the data frame

In [25]:
rec_cases_df = covid_df[['date', 'daily_recovered']]
rec_cases_df

date  daily_recovered
0     30 January                 0
1     31 January                 0
2    01 February                 0
3    02 February                 0
4    03 February                 0
..            ...              ...
336  31 December             21969
337   01 January             23838
338   02 January             20903
339   03 January             19658
340   04 January             29209

[341 rows x 2 columns]

The new data frame 'rec_cases_df' is simply a 'snapshot' of the original data frame covid_df. Both data frames point to the same location in computer memory which effectively means that changing values in one date frame will affect the other data frame as well. This sharing of data between data frames makes data manipulation really fast in Pandas.

Sometimes though we may need to make a copy of the data frame, which can be done using the 'copy()' method

In [26]:
covid_df_copy = covid_df.copy()

In this case, the covid_df_copy data frame is separate from covid_df, and so changing values in the copy will not affect the original data frame

To access a specific row in the data, Pandas provides the 'loc' method

In [27]:
covid_df.loc[250]

date               06 October 
date_ymd            2020-10-06
daily_confirmed          71869
total_confirmed        6754191
daily_recovered          81945
total_recovered        5741231
daily_deceased             990
total_deceased          104005
Name: 250, dtype: object

We had earlier seen the 'head()' method that gives us the first 5 rows from the data frame. We can pass in a number as an argument to the method to view that many rows

In [28]:
covid_df.head(20)

date    date_ymd  daily_confirmed  total_confirmed  \
0    30 January   2020-01-30                1                1   
1    31 January   2020-01-31                0                1   
2   01 February   2020-02-01                0                1   
3   02 February   2020-02-02                1                2   
4   03 February   2020-02-03                1                3   
5   04 February   2020-02-04                0                3   
6   05 February   2020-02-05                0                3   
7   06 February   2020-02-06                0                3   
8   07 February   2020-02-07                0                3   
9   08 February   2020-02-08                0                3   
10  09 February   2020-02-09                0                3   
11  10 February   2020-02-10                0                3   
12  11 February   2020-02-11                0                3   
13  12 February   2020-02-12                0                3   
14  13 February   2020-02-13                0                3   
15  14 February   2020-02-14                0                3   
16  15 February   2020-02-15                0                3   
17  16 February   2020-02-16                0                3   
18  17 February   2020-02-17                0                3   
19  18 February   2020-02-18                0                3   

    daily_recovered  total_recovered  daily_deceased  total_deceased  
0                 0                0               0               0  
1                 0                0               0               0  
2                 0                0               0               0  
3                 0                0               0               0  
4                 0                0               0               0  
5                 0                0               0               0  
6                 0                0               0               0  
7                 0                0               0               0  
8                 0                0               0               0  
9                 0                0               0               0  
10                0                0               0               0  
11                0                0               0               0  
12                0                0               0               0  
13                0                0               0               0  
14                1                1               0               0  
15                0                1               0               0  
16                0                1               0               0  
17                1                2               0               0  
18                0                2               0               0  
19                0                2               0               0

The 'tail()' method shows us the last 5 rows

In [29]:
covid_df.tail()

date    date_ymd  daily_confirmed  total_confirmed  \
336  31 December   2020-12-31            19026         10286234   
337   01 January   2021-01-01            20159         10306393   
338   02 January   2021-01-02            18144         10324537   
339   03 January   2021-01-03            16678         10341215   
340   04 January   2021-01-04            16278         10357493   

     daily_recovered  total_recovered  daily_deceased  total_deceased  
336            21969          9881565             244          148427  
337            23838          9905403             237          148664  
338            20903          9926306             216          148880  
339            19658          9945964             215          149095  
340            29209          9975173             200          149295

We an use the 'sample()' method to retrieve a random sample of rows from the data

In [30]:
covid_df.sample(10)

date    date_ymd  daily_confirmed  total_confirmed  \
158       06 July   2020-07-06            22500           720349   
218  04 September   2020-09-04            87115          4020252   
275    31 October   2020-10-31            47228          8183317   
288   13 November   2020-11-13            44620          8773167   
328   23 December   2020-12-23            24716         10123948   
124       02 June   2020-06-02             8812           207187   
118        27 May   2020-05-27             7246           158104   
44       14 March   2020-03-14               11              102   
43       13 March   2020-03-13               10               91   
57       27 March   2020-03-27              153              883   

     daily_recovered  total_recovered  daily_deceased  total_deceased  
158            15315           440210             473           20176  
218            69625          3104505             950           69179  
275            58682          7489426             469          121559  
288            47620          8161355             517          128634  
328            29946          9692476             315          146200  
124             4531           100275             222            5829  
118             3434            67725             188            4534  
44                 0               10               1               2  
43                 6               10               0               1  
57                25               75               3              19

We can check if our data frame contains any null / missing values by using the 'isnull().sum()' expression

In [31]:
covid_df.isnull().sum()

date               0
date_ymd           0
daily_confirmed    0
total_confirmed    0
daily_recovered    0
total_recovered    0
daily_deceased     0
total_deceased     0
dtype: int64

Here's a summary of what we learnt in the above section:
    - covid_df['daily_recovered']: retrieving columns as a series using the column name
    - daily_recovered[250]: retrieving values from a series using an index
    - covid_df.at[250, 'daily_recovered]: retrieving a single value from the data frame
    - covid_df.copy(): create a deep copy of the data frame
    - covid_df.loc[250]: retrieving a row or range of rows from the data frame
    - head, tail and sample: methods used to retrieve several rows from the data frame